src/test/scala/org/broadinstitute/dsde/firecloud/webservice/ImportPermissionApiServiceSpec.scala

```diff
@@ -0,0 +1,157 @@
+        RawlsBillingProjectMembership(RawlsBillingProjectName("projectone"), ProjectRoles.User, CreationStatuses.Ready, None),
+        RawlsBillingProjectMembership(RawlsBillingProjectName("projecttwo"), ProjectRoles.Owner, CreationStatuses.Creating, None)
+      ))
+      case "projectsNotReady" => Future.successful(Seq(
+        RawlsBillingProjectMembership(RawlsBillingProjectName("projectone"), ProjectRoles.User, CreationStatuses.Creating, None),
+        RawlsBillingProjectMembership(RawlsBillingProjectName("projecttwo"), ProjectRoles.Owner, CreationStatuses.Creating, None)
+
+      ))
+      case "noProjects" => Future.successful(Seq.empty[RawlsBillingProjectMembership])
+      case _ => Future.failed(new FireCloudException("intentional exception for getProjects catchall case"))
+    }
+  }
+
+  override def getWorkspaces(implicit userInfo: WithAccessToken): Future[Seq[WorkspaceListResponse]] = {
+    parseTestToken(userInfo)._1 match {
+      case "hasWorkspaces" => Future.successful(Seq(
+        WorkspaceListResponse(WorkspaceAccessLevels.ProjectOwner, newWorkspace, WorkspaceSubmissionStats(None, None, runningSubmissionsCount = 0), List.empty, Some(false)),
+        WorkspaceListResponse(WorkspaceAccessLevels.Read, newWorkspace, WorkspaceSubmissionStats(None, None, runningSubmissionsCount = 0), List.empty, Some(false)),
+        WorkspaceListResponse(WorkspaceAccessLevels.Owner, publishedRawlsWorkspaceWithAttributes, WorkspaceSubmissionStats(None, None, runningSubmissionsCount = 0), List.empty, Some(false)),
+        WorkspaceListResponse(WorkspaceAccessLevels.NoAccess, newWorkspace, WorkspaceSubmissionStats(None, None, runningSubmissionsCount = 0), List.empty, Some(false))
+      ))
+      case "onlyReadableWorkspaces" => Future.successful(Seq(
+        WorkspaceListResponse(WorkspaceAccessLevels.Read, newWorkspace, WorkspaceSubmissionStats(None, None, runningSubmissionsCount = 0), List.empty, Some(false)),
+        WorkspaceListResponse(WorkspaceAccessLevels.Read, newWorkspace, WorkspaceSubmissionStats(None, None, runningSubmissionsCount = 0), List.empty, Some(false)),
+        WorkspaceListResponse(WorkspaceAccessLevels.Read, publishedRawlsWorkspaceWithAttributes, WorkspaceSubmissionStats(None, None, runningSubmissionsCount = 0), List.empty, Some(false)),
+        WorkspaceListResponse(WorkspaceAccessLevels.NoAccess, newWorkspace, WorkspaceSubmissionStats(None, None, runningSubmissionsCount = 0), List.empty, Some(false))
+      ))
+      case "noWorkspaces" => Future.successful(Seq.empty[WorkspaceListResponse])
+      case _ => Future.failed(new FireCloudException("intentional exception for getWorkspaces catchall case"))
+    }
+  }
+
+  // this is hacky, but the only argument to getProjects and getWorkspaces is the access token. Therefore,
+  // we need to encode our test criteria into a string, and we can do so using a delimeter.
+  private def parseTestToken(userInfo: WithAccessToken): (String,String) = {
```

Comment:
Yes, this is an interesting approach (👍). The other way I can think of is to create completely separate test classes (each with a custom override) for each of the 4 cases ... but that seems like far more work than necessary.